### ISSUE TITLE
Type inference issue JSX with generic

### ISSUE THREAD

**Issue by ulrichb** [Microsoft/TypeScript#30057]
**TypeScript Version:**  3.3.3333

This is a follow-up for #27986 (see also [this comment](https://github.com/Microsoft/TypeScript/issues/27986#issuecomment-466425191)).  It already has been fixed and the type inference worked fine since then. But now, the exact same compiler error (see original issue for code snippet) appeared after updating `@types/react`:

![image](https://user-images.githubusercontent.com/388796/53250094-00fe6600-36ba-11e9-84e2-011d8e86b3a3.png)

I've tacked it down and it happened exactly with `@types/react@16.7.21` (version `@types/react@16.7.20` and before works).

I've generated a diff of these two versions: [Compare.pdf](https://github.com/Microsoft/TypeScript/files/2894517/Compare.pdf)

And here are the full packages: [both_versions.zip](https://github.com/Microsoft/TypeScript/files/2894521/both_versions.zip)

**Comment by weswigham** [comment]
I'm not seeing the same error when I try to make a repro - what are you compiler options?

**Comment by weswigham** [comment]
Actually, hold on (well, maybe in addition to that question), that looks like VS - what version of the typescript SDK do you have installed in VS? IIRC the TS version installed in your project by `npm` has no bearing on the version used by VS for editor features.

**Comment by ulrichb** [comment]
Hi. Also happens via `tsc`. Here is the command line.

```plain
>npx tsc -t es2017 -m commonjs --jsx react --strict TypeScriptIssue27986.tsx
TypeScriptIssue27986.tsx:24:26 - error TS2322: Type '() => Promise<ErrorResult | { success: true; }>' is not assignable to type '() => Promise<{ success: true; }>'.
  Type 'Promise<ErrorResult | { success: true; }>' is not assignable to type 'Promise<{ success: true; }>'.
    Type 'ErrorResult | { success: true; }' is not assignable to type '{ success: true; }'.
      Property 'success' is missing in type 'ErrorResult' but required in type '{ success: true; }'.

24             <AsyncLoader asyncLoad={load}>{result =>
                            ~~~~~~~~~

  TypeScriptIssue27986.tsx:16:34
    16 async function load(): Promise<{ success: true } | ErrorResult> {
                                        ~~~~~~~
    'success' is declared here.
  TypeScriptIssue27986.tsx:8:14
    8     readonly asyncLoad: () => Promise<TResult>;
                   ~~~~~~~~~
    The expected type comes from property 'asyncLoad' which is declared here on type 'IntrinsicAttributes & IntrinsicClassAttributes<AsyncLoader<{ success: true; }>> & Readonly<{ children?: ReactNode; }> & Readonly<AsyncLoaderProps<{ success: true; }>>'


Found 1 error.

>npx tsc -v
Version 3.3.3333
```

Same with TSC 3.4.0-dev.20190222.

**Comment by ulrichb** [comment]
(As in #27986 `--strict` is necessary.)

**Comment by ulrichb** [comment]
Oh, that's interesting: My ["more simple repro"](https://github.com/Microsoft/TypeScript/issues/27986#issuecomment-431484440) from the comments _does not_ emit an error. (Only the full sample in the issue description.)

**Comment by ulrichb** [comment]
@weswigham Did you have a look on this? Atm. this prevents me from upgrading to React 16.8 (typings).

**Comment by weswigham** [comment]
Minimal repro without involving `react`:
```ts
// @strict: true
// @jsx: react
// @lib: es6

interface ErrorResult {
    readonly error: true
}

interface AsyncLoaderProps<TResult extends {}> {
    readonly asyncLoad: () => Promise<TResult>;
    readonly children: (result: Exclude<TResult, ErrorResult>) => string;
}

class AsyncLoader<TResult extends {}> {
    constructor(props: Readonly<AsyncLoaderProps<TResult>>);
    constructor(props: AsyncLoaderProps<TResult>, context?: any);
    constructor(...args: any[]) {}
}

async function load(): Promise<{ success: true } | ErrorResult> {
    return { success: true };
}

new AsyncLoader({
    asyncLoad: load,
    children: result => result.success as any,
});
```
eliminating either `constructor` overload removes the error- which means you can work around the bug in your own code by writing your own constructor out explicitly in your `Component` subtype.

**Comment by ulrichb** [comment]
Many thanks for having a look. And thanks for the hint with the c'tor workaround!

**Comment by ulrichb** [comment]
@weswigham Thank you very much for the fix!

**Comment by city41** [comment]
I believe I am seeing this issue with TS 3.4.5

This file fails to compile:

```
import React, { ErrorInfo } from 'react';

interface IErrorProps {
    error: Error;
    errorInfo: ErrorInfo;
}

interface IState {
    caughtError: boolean;
    error: Error | null;
    errorInfo: ErrorInfo | null;
}

export default function withErrorBoundary<TErrorProps extends IErrorProps, TComponentProps>(
    Component: React.ComponentType<TComponentProps>,
    ErrorCmp: React.ComponentType<TErrorProps>
) {
    return class extends React.PureComponent<TComponentProps, IState> {
        state: IState = {
            caughtError: false,
            error: null,
            errorInfo: null
        };

        componentDidCatch(error: Error, errorInfo: ErrorInfo) {
            this.setState({
                caughtError: true,
                error,
                errorInfo
            });
        }

        render() {
            const { caughtError, error, errorInfo } = this.state;

            if (caughtError && error && errorInfo) {
                return <ErrorCmp error={error} errorInfo={errorInfo} />;
            } else {
                return <Component {...this.props} />;
            }
        }
    }
}
```

with the error `Type '{ error: Error; errorInfo: ErrorInfo; }' is not assignable to type 'TErrorProps'`

But if I downgrade to TS 3.1.4, it compiles without issue.

**Comment by city41** [comment]
My above comment is not correct, it looks like I am seeing an improvement/bug fix from TS 3.1 to 3.2, more info here: https://stackoverflow.com/questions/55942402/when-using-generic-react-components-get-error-in-ts-3-2
